### PR TITLE
Add some benchmarks

### PR DESCRIPTION
**Pull request by dangleptr**
1.  Benchmark for std::transform VS range for
2.  Benchmark for std::for_each  VS range for
3.  Benchmark for MultiVersion VS SingleVersion in rocksdb

### CODE REVIEW

**Comment by dangleptr** [comment]
Jenkins go

File: src/common/base/test/RangeVsTransformBenchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+BENCHMARK(Test1_RangeTestStr) {
+    std::vector<int32_t> from;
+    std::vector<std::string> to;
+    BENCHMARK_SUSPEND {
+        from.resize(1000, 0);
+        to.resize(1000);
+    }
+    int32_t index = 0;
+    for (auto& i : from) {
+        to[index++] = std::to_string(i);
```
Comment (dutor) [inline src/common/base/test/RangeVsTransformBenchmark.cpp]: You are testing `std::to_string`, not `std::transform` vs. range for.

File: src/common/base/test/RangeVsTransformBenchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+-O3
+[heng.chen@hp-server nebula]$ src/common/base/test/_build/range_vs_transform__bm
+============================================================================
+/data/heng/workspace/nebula/src/common/base/test/RangeVsTransformBenchmark.cpprelative  time/iter  iters/s
+============================================================================
+Test1_RangeTestStr                                          84.19us   11.88K
+Test1_TransformStr                               100.01%    84.18us   11.88K
+----------------------------------------------------------------------------
+Test2_RangeTestInt                                         222.70ns    4.49M
```
Comment (dutor) [inline src/common/base/test/RangeVsTransformBenchmark.cpp]: Please expain why `-O3` is better than `-O2` for this case.

File: src/common/base/test/CMakeLists.txt [hunk 3]
```diff
@@ -110,3 +110,11 @@ nebula_link_libraries(
     gtest_main
 )
 nebula_add_test(nebulakey_utils_test)
+
+add_compile_options(-O3)
```
Comment (dutor) [inline src/common/base/test/CMakeLists.txt]: Please use `target_compile_options` instead.

File: src/common/base/test/RangeVsTransformBenchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+int main(int argc, char** argv) {
+    folly::init(&argc, &argv, true);
+    folly::runBenchmarks();
+    return 0;
+}
+
+
+/*
+-O2
+[heng.chen@hp-server nebula]$ src/common/base/test/_build/range_vs_transform__bm
```
Comment (dutor) [inline src/common/base/test/RangeVsTransformBenchmark.cpp]: Better no to expose such kind of privacy.

Better post the CPU model you are testing on.

File: src/common/base/test/RangeVsTransformBenchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+============================================================================
+/data/heng/workspace/nebula/src/common/base/test/RangeVsTransformBenchmark.cpprelative  time/iter  iters/s
```
Comment (dutor) [inline src/common/base/test/RangeVsTransformBenchmark.cpp]: ditto. 

Besides, try to align the headers according to the testing results

File: src/common/base/test/RangeVsTransformBenchmark.cpp (hunk 1 above)
Comment (dangleptr) [reply src/common/base/test/RangeVsTransformBenchmark.cpp]: int to std::string is a common case with std::transform.   We'd better to figure out the accurate perf under this case.

File: src/common/base/test/RangeVsTransformBenchmark.cpp (hunk 2 above)
Comment (dangleptr) [reply src/common/base/test/RangeVsTransformBenchmark.cpp]: Let me try it.

Comment (dangleptr) [reply src/common/base/test/RangeVsTransformBenchmark.cpp]: Generate the asm for the snippet
```
int from[100000];
int to[100000];

int f() {
    int index = 0;
    for (auto& i : from) {
        to[index++] = i;
    }
    return 0;
}
```
With -O2
```
.LFB1898:
	.cfi_startproc
	xorl	%eax, %eax
	.p2align 4,,10
	.p2align 3
.L2:
	movl	from(%rax), %edx
	addq	$4, %rax
	movl	%edx, to-4(%rax)
	cmpq	$400000, %rax
	jne	.L2
	xorl	%eax, %eax
	ret
	.cfi_endproc
```
With -O3
```
_Z1fv:
.LFB1898:
	.cfi_startproc
	subq	$8, %rsp
	.cfi_def_cfa_offset 16
	movl	$400000, %edx
	movl	$from, %esi
	movl	$to, %edi
	call	memcpy
	xorl	%eax, %eax
	addq	$8, %rsp
	.cfi_def_cfa_offset 8
	ret
	.cfi_endproc
```

Notice that,  with O3,  there is no loop,  just call memcpy directly,  very smart!

File: src/common/base/test/RangeVsTransformBenchmark.cpp (hunk 5 above)
Comment (dutor) [reply src/common/base/test/RangeVsTransformBenchmark.cpp]: Actually, I had digged into the implementations of follybenchmark, trying to make it able to use relative paths, but ended up with no solution.

File: src/common/base/test/RangeVsTransformBenchmark.cpp (hunk 2 above)
Comment (dutor) [reply src/common/base/test/RangeVsTransformBenchmark.cpp]: Excellent.

File: src/common/base/test/RangeVsTransformBenchmark.cpp (hunk 1 above)
Comment (dutor) [reply src/common/base/test/RangeVsTransformBenchmark.cpp]: Fair enough.

File: src/common/base/test/CMakeLists.txt [hunk 6]
```diff
@@ -110,3 +110,11 @@ nebula_link_libraries(
     gtest_main
 )
 nebula_add_test(nebulakey_utils_test)
+
+add_executable(range_vs_transform__bm RangeVsTransformBenchmark.cpp)
+nebula_link_libraries(
+    range_vs_transform__bm
```
Comment (darionyaphet) [inline src/common/base/test/CMakeLists.txt]: one more underscore

File: src/kvstore/test/MultiVersionBenchmark.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        if (FLAGS_do_compact) {
+            rocksdb::CompactRangeOptions croptions;
+            db->CompactRange(croptions, nullptr, nullptr);
+        }
+    }
+    auto start = genKey(0, 0);
+    rocksdb::ReadOptions roptions;
+    rocksdb::Iterator* iter = db->NewIterator(roptions);
+    if (iter) {
+        iter->Seek(rocksdb::Slice(start));
```
Comment (darionyaphet) [inline src/kvstore/test/MultiVersionBenchmark.cpp]: could  std::move here ?

File: src/kvstore/test/MultiVersionBenchmark.cpp [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+            for (int v = 0; v < FLAGS_versions; v++) {
+                std::string key;
+                if (withVersion) {
+                    key = genKey(i, v);
+                } else {
+                    key = genKey(i, 0);
+                }
+                auto val = folly::stringPrintf("val_%d_%d", i, v);
+                db->Put(woptions,
+                        rocksdb::Slice(key.data(), key.size()),
```
Comment (darionyaphet) [inline src/kvstore/test/MultiVersionBenchmark.cpp]: should std::move key and val  ?

File: src/common/base/test/CMakeLists.txt [hunk 9]
```diff
@@ -110,3 +110,11 @@ nebula_link_libraries(
...
+    follybenchmark
+    boost_regex
```
Comment (darionyaphet) [inline src/common/base/test/CMakeLists.txt]: will use `boost_regex` ?

File: src/kvstore/test/MultiVersionBenchmark.cpp (hunk 7 above)
Comment (dangleptr) [reply src/kvstore/test/MultiVersionBenchmark.cpp]: For slice,  move equals copy

File: src/kvstore/test/MultiVersionBenchmark.cpp (hunk 8 above)
Comment (dangleptr) [reply src/kvstore/test/MultiVersionBenchmark.cpp]: ditto

File: src/common/base/test/CMakeLists.txt (hunk 9 above)
Comment (dangleptr) [reply src/common/base/test/CMakeLists.txt]: because it is depended by folly benchmark

File: src/common/base/test/CMakeLists.txt (hunk 6 above)
Comment (dangleptr) [reply src/common/base/test/CMakeLists.txt]: Good catch

**Comment by dangleptr** [comment]
Jenkins go

File: src/kvstore/test/MultiVersionBenchmark.cpp (hunk 7 above)
Comment (dutor) [reply src/kvstore/test/MultiVersionBenchmark.cpp]: > For slice, move equals copy

FYI. This also appies to StringPiece.

**Pull request merged**